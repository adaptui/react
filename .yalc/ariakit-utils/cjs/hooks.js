'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('react');
var dom = require('./dom.js');
var misc = require('./misc.js');

const useReactId = typeof react.useId === "function" ? react.useId : undefined;
const useReactDeferredValue = typeof react.useDeferredValue === "function" ? react.useDeferredValue : undefined;
/**
 * `React.useLayoutEffect` that fallbacks to `React.useEffect` on server side.
 */

const useSafeLayoutEffect = dom.canUseDOM ? react.useLayoutEffect : react.useEffect;
/**
 * Returns a value that never changes even if the argument is updated.
 * @example
 * function Component({ prop }) {
 *   const initialProp = useInitialValue(prop);
 * }
 */

function useInitialValue(value) {
  const [initialValue] = react.useState(value);
  return initialValue;
}
/**
 * Returns a value that is lazily initiated and never changes.
 * @example
 * function Component() {
 *   const set = useLazyRef(() => new Set());
 * }
 */

function useLazyRef(init) {
  const ref = react.useRef();

  if (ref.current === undefined) {
    ref.current = init();
  }

  return ref.current;
}
/**
 * Creates a `React.RefObject` that is constantly updated with the incoming
 * value.
 * @example
 * function Component({ prop }) {
 *   const propRef = useLiveRef(prop);
 * }
 */

function useLiveRef(value) {
  const ref = react.useRef(value);
  useSafeLayoutEffect(() => {
    ref.current = value;
  });
  return ref;
}
/**
 * Creates a memoized callback function that is constantly updated with the
 * incoming callback.
 * @example
 * function Component(props) {
 *   const onClick = useEventCallback(props.onClick);
 *   React.useEffect(() => {}, [onClick]);
 * }
 */

function useEventCallback(callback) {
  // @ts-ignore
  const ref = react.useRef(() => {
    throw new Error("Cannot call an event handler while rendering.");
  });
  useSafeLayoutEffect(() => {
    ref.current = callback;
  });
  return react.useCallback( // @ts-ignore
  function () {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return ref.current == null ? void 0 : ref.current(...args);
  }, []);
}
/**
 * Merges React Refs into a single memoized function ref so you can pass it to
 * an element.
 * @example
 * const Component = React.forwardRef((props, ref) => {
 *   const internalRef = React.useRef();
 *   return <div {...props} ref={useForkRef(internalRef, ref)} />;
 * });
 */

function useForkRef() {
  for (var _len2 = arguments.length, refs = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    refs[_key2] = arguments[_key2];
  }

  return react.useMemo(() => {
    if (!refs.some(Boolean)) return;
    return value => {
      refs.forEach(ref => {
        misc.setRef(ref, value);
      });
    };
  }, refs);
}
/**
 * Returns the ref element's ID.
 */

function useRefId(ref, deps) {
  const [id, setId] = react.useState(undefined);
  useSafeLayoutEffect(() => {
    var _ref$current;

    setId(ref == null ? void 0 : (_ref$current = ref.current) == null ? void 0 : _ref$current.id);
  }, deps);
  return id;
}
/**
 * Generates a unique ID. Uses React's useId if available.
 */

function useId(defaultId) {
  if (useReactId) {
    const reactId = useReactId();
    if (defaultId) return defaultId;
    return reactId;
  }

  const [id, setId] = react.useState(defaultId);
  useSafeLayoutEffect(() => {
    if (defaultId || id) return;
    const random = Math.random().toString(36).substr(2, 6);
    setId("id-" + random);
  }, [defaultId, id]);
  return defaultId || id;
}
/**
 * Uses React's useDeferredValue if available.
 */

function useDeferredValue(value) {
  if (useReactDeferredValue) {
    return useReactDeferredValue(value);
  }

  const [deferredValue, setDeferredValue] = react.useState(value);
  react.useEffect(() => {
    const raf = requestAnimationFrame(() => {
      setDeferredValue(value);
    });
    return () => cancelAnimationFrame(raf);
  }, [value]);
  return deferredValue;
}
/**
 * Returns the tag name by parsing an element ref and the `as` prop.
 * @example
 * function Component(props) {
 *   const ref = React.useRef();
 *   const tagName = useTagName(ref, "button"); // div
 *   return <div ref={ref} {...props} />;
 * }
 */

function useTagName(ref, type) {
  const [tagName, setTagName] = react.useState(() => stringOrUndefined(type));
  useSafeLayoutEffect(() => {
    var _ref$current2;

    setTagName((ref == null ? void 0 : (_ref$current2 = ref.current) == null ? void 0 : _ref$current2.tagName.toLowerCase()) || stringOrUndefined(type));
  }, [ref, type]);
  return tagName;
}

function stringOrUndefined(type) {
  if (typeof type === "string") {
    return type;
  }

  return;
}
/**
 * A `React.useEffect` that will not run on the first render.
 */


function useUpdateEffect(effect, deps) {
  const mounted = react.useRef(false);
  react.useEffect(() => {
    if (mounted.current) {
      return effect();
    }

    mounted.current = true;
  }, deps);
}
/**
 * A `React.useLayoutEffect` that will not run on the first render.
 */

function useUpdateLayoutEffect(effect, deps) {
  const mounted = react.useRef(false);
  useSafeLayoutEffect(() => {
    if (mounted.current) {
      return effect();
    }

    mounted.current = true;
  }, deps);
}
/**
 * A custom version of `React.useState` that uses the `state` and `setState`
 * arguments. If they're not provided, it will use the internal state.
 */

function useControlledState(defaultState, state, setState) {
  const [localState, setLocalState] = react.useState(defaultState);
  const nextState = state !== undefined ? state : localState;
  const stateRef = useLiveRef(state);
  const setStateRef = useLiveRef(setState);
  const nextStateRef = useLiveRef(nextState);
  const setNextState = react.useCallback(prevValue => {
    const setStateProp = setStateRef.current;

    if (setStateProp) {
      if (isSetNextState(setStateProp)) {
        setStateProp(prevValue);
      } else {
        const nextValue = misc.applyState(prevValue, nextStateRef.current);
        nextStateRef.current = nextValue;
        setStateProp(nextValue);
      }
    }

    if (stateRef.current === undefined) {
      setLocalState(prevValue);
    }
  }, []);
  defineSetNextState(setNextState);
  return [nextState, setNextState];
}
const SET_NEXT_STATE = Symbol("setNextState");

function isSetNextState(arg) {
  return arg[SET_NEXT_STATE] === true;
}

function defineSetNextState(arg) {
  if (!isSetNextState(arg)) {
    Object.defineProperty(arg, SET_NEXT_STATE, {
      value: true
    });
  }
}
/**
 * A React hook similar to `useState` and `useReducer`, but with the only
 * purpose of re-rendering the component.
 */


function useForceUpdate() {
  return react.useReducer(() => ({}), {});
}
/**
 * Returns an event callback similar to `useEventCallback`, but this also
 * accepts a boolean value, which will be turned into a function.
 */

function useBooleanEventCallback(booleanOrCallback) {
  return useEventCallback(typeof booleanOrCallback === "function" ? booleanOrCallback : () => booleanOrCallback);
}
/**
 * Returns props with an additional `wrapElement` prop.
 */

function useWrapElement(props, callback, deps) {
  if (deps === void 0) {
    deps = [];
  }

  const wrapElement = react.useCallback(element => {
    if (props.wrapElement) {
      element = props.wrapElement(element);
    }

    return callback(element);
  }, [...deps, props.wrapElement]);
  return { ...props,
    wrapElement
  };
}

exports.useBooleanEventCallback = useBooleanEventCallback;
exports.useControlledState = useControlledState;
exports.useDeferredValue = useDeferredValue;
exports.useEventCallback = useEventCallback;
exports.useForceUpdate = useForceUpdate;
exports.useForkRef = useForkRef;
exports.useId = useId;
exports.useInitialValue = useInitialValue;
exports.useLazyRef = useLazyRef;
exports.useLiveRef = useLiveRef;
exports.useRefId = useRefId;
exports.useSafeLayoutEffect = useSafeLayoutEffect;
exports.useTagName = useTagName;
exports.useUpdateEffect = useUpdateEffect;
exports.useUpdateLayoutEffect = useUpdateLayoutEffect;
exports.useWrapElement = useWrapElement;
