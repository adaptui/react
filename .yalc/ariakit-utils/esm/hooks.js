import { useLayoutEffect, useEffect, useState, useRef, useCallback, useMemo, useReducer, useId as useId$1, useDeferredValue as useDeferredValue$1 } from 'react';
import { canUseDOM } from './dom.js';
import { setRef, applyState } from './misc.js';

const useReactId = typeof useId$1 === "function" ? useId$1 : undefined;
const useReactDeferredValue = typeof useDeferredValue$1 === "function" ? useDeferredValue$1 : undefined;
/**
 * `React.useLayoutEffect` that fallbacks to `React.useEffect` on server side.
 */

const useSafeLayoutEffect = canUseDOM ? useLayoutEffect : useEffect;
/**
 * Returns a value that never changes even if the argument is updated.
 * @example
 * function Component({ prop }) {
 *   const initialProp = useInitialValue(prop);
 * }
 */

function useInitialValue(value) {
  const [initialValue] = useState(value);
  return initialValue;
}
/**
 * Returns a value that is lazily initiated and never changes.
 * @example
 * function Component() {
 *   const set = useLazyRef(() => new Set());
 * }
 */

function useLazyRef(init) {
  const ref = useRef();

  if (ref.current === undefined) {
    ref.current = init();
  }

  return ref.current;
}
/**
 * Creates a `React.RefObject` that is constantly updated with the incoming
 * value.
 * @example
 * function Component({ prop }) {
 *   const propRef = useLiveRef(prop);
 * }
 */

function useLiveRef(value) {
  const ref = useRef(value);
  useSafeLayoutEffect(() => {
    ref.current = value;
  });
  return ref;
}
/**
 * Creates a memoized callback function that is constantly updated with the
 * incoming callback.
 * @example
 * function Component(props) {
 *   const onClick = useEventCallback(props.onClick);
 *   React.useEffect(() => {}, [onClick]);
 * }
 */

function useEventCallback(callback) {
  // @ts-ignore
  const ref = useRef(() => {
    throw new Error("Cannot call an event handler while rendering.");
  });
  useSafeLayoutEffect(() => {
    ref.current = callback;
  });
  return useCallback( // @ts-ignore
  function () {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return ref.current == null ? void 0 : ref.current(...args);
  }, []);
}
/**
 * Merges React Refs into a single memoized function ref so you can pass it to
 * an element.
 * @example
 * const Component = React.forwardRef((props, ref) => {
 *   const internalRef = React.useRef();
 *   return <div {...props} ref={useForkRef(internalRef, ref)} />;
 * });
 */

function useForkRef() {
  for (var _len2 = arguments.length, refs = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    refs[_key2] = arguments[_key2];
  }

  return useMemo(() => {
    if (!refs.some(Boolean)) return;
    return value => {
      refs.forEach(ref => {
        setRef(ref, value);
      });
    };
  }, refs);
}
/**
 * Returns the ref element's ID.
 */

function useRefId(ref, deps) {
  const [id, setId] = useState(undefined);
  useSafeLayoutEffect(() => {
    var _ref$current;

    setId(ref == null ? void 0 : (_ref$current = ref.current) == null ? void 0 : _ref$current.id);
  }, deps);
  return id;
}
/**
 * Generates a unique ID. Uses React's useId if available.
 */

function useId(defaultId) {
  if (useReactId) {
    const reactId = useReactId();
    if (defaultId) return defaultId;
    return reactId;
  }

  const [id, setId] = useState(defaultId);
  useSafeLayoutEffect(() => {
    if (defaultId || id) return;
    const random = Math.random().toString(36).substr(2, 6);
    setId("id-" + random);
  }, [defaultId, id]);
  return defaultId || id;
}
/**
 * Uses React's useDeferredValue if available.
 */

function useDeferredValue(value) {
  if (useReactDeferredValue) {
    return useReactDeferredValue(value);
  }

  const [deferredValue, setDeferredValue] = useState(value);
  useEffect(() => {
    const raf = requestAnimationFrame(() => {
      setDeferredValue(value);
    });
    return () => cancelAnimationFrame(raf);
  }, [value]);
  return deferredValue;
}
/**
 * Returns the tag name by parsing an element ref and the `as` prop.
 * @example
 * function Component(props) {
 *   const ref = React.useRef();
 *   const tagName = useTagName(ref, "button"); // div
 *   return <div ref={ref} {...props} />;
 * }
 */

function useTagName(ref, type) {
  const [tagName, setTagName] = useState(() => stringOrUndefined(type));
  useSafeLayoutEffect(() => {
    var _ref$current2;

    setTagName((ref == null ? void 0 : (_ref$current2 = ref.current) == null ? void 0 : _ref$current2.tagName.toLowerCase()) || stringOrUndefined(type));
  }, [ref, type]);
  return tagName;
}

function stringOrUndefined(type) {
  if (typeof type === "string") {
    return type;
  }

  return;
}
/**
 * A `React.useEffect` that will not run on the first render.
 */


function useUpdateEffect(effect, deps) {
  const mounted = useRef(false);
  useEffect(() => {
    if (mounted.current) {
      return effect();
    }

    mounted.current = true;
  }, deps);
}
/**
 * A `React.useLayoutEffect` that will not run on the first render.
 */

function useUpdateLayoutEffect(effect, deps) {
  const mounted = useRef(false);
  useSafeLayoutEffect(() => {
    if (mounted.current) {
      return effect();
    }

    mounted.current = true;
  }, deps);
}
/**
 * A custom version of `React.useState` that uses the `state` and `setState`
 * arguments. If they're not provided, it will use the internal state.
 */

function useControlledState(defaultState, state, setState) {
  const [localState, setLocalState] = useState(defaultState);
  const nextState = state !== undefined ? state : localState;
  const stateRef = useLiveRef(state);
  const setStateRef = useLiveRef(setState);
  const nextStateRef = useLiveRef(nextState);
  const setNextState = useCallback(prevValue => {
    const setStateProp = setStateRef.current;

    if (setStateProp) {
      if (isSetNextState(setStateProp)) {
        setStateProp(prevValue);
      } else {
        const nextValue = applyState(prevValue, nextStateRef.current);
        nextStateRef.current = nextValue;
        setStateProp(nextValue);
      }
    }

    if (stateRef.current === undefined) {
      setLocalState(prevValue);
    }
  }, []);
  defineSetNextState(setNextState);
  return [nextState, setNextState];
}
const SET_NEXT_STATE = Symbol("setNextState");

function isSetNextState(arg) {
  return arg[SET_NEXT_STATE] === true;
}

function defineSetNextState(arg) {
  if (!isSetNextState(arg)) {
    Object.defineProperty(arg, SET_NEXT_STATE, {
      value: true
    });
  }
}
/**
 * A React hook similar to `useState` and `useReducer`, but with the only
 * purpose of re-rendering the component.
 */


function useForceUpdate() {
  return useReducer(() => ({}), {});
}
/**
 * Returns an event callback similar to `useEventCallback`, but this also
 * accepts a boolean value, which will be turned into a function.
 */

function useBooleanEventCallback(booleanOrCallback) {
  return useEventCallback(typeof booleanOrCallback === "function" ? booleanOrCallback : () => booleanOrCallback);
}
/**
 * Returns props with an additional `wrapElement` prop.
 */

function useWrapElement(props, callback, deps) {
  if (deps === void 0) {
    deps = [];
  }

  const wrapElement = useCallback(element => {
    if (props.wrapElement) {
      element = props.wrapElement(element);
    }

    return callback(element);
  }, [...deps, props.wrapElement]);
  return { ...props,
    wrapElement
  };
}

export { useBooleanEventCallback, useControlledState, useDeferredValue, useEventCallback, useForceUpdate, useForkRef, useId, useInitialValue, useLazyRef, useLiveRef, useRefId, useSafeLayoutEffect, useTagName, useUpdateEffect, useUpdateLayoutEffect, useWrapElement };
