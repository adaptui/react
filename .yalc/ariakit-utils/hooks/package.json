{
  "name": "ariakit-utils/hooks",
  "private": true,
  "sideEffects": false,
  "main": "../cjs/hooks",
  "module": "../esm/hooks",
  "types": "../ts/hooks.d.ts"
}