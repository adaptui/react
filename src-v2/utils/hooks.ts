import { useEffect, useRef } from "react";

/**
 * usePrevious hook for React
 *
 * @param currentValue The value whose previous value is to be tracked
 * @returns The previous value
 */
export function usePrevious<T>(currentValue: T): T | null {
  const previousRef = useRef<T | null>(null);

  useEffect(() => {
    previousRef.current = currentValue;
  });

  return previousRef.current;
}
