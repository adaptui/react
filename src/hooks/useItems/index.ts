export * from "./useItems";
export * from "./types";
