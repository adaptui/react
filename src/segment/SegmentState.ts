/**
 * All credit goes to [React Spectrum](https://github.com/adobe/react-spectrum)
 * We improved the Calendar from Stately [useCalendarState](https://github.com/adobe/react-spectrum/tree/main/packages/%40react-stately/calendar)
 * to work with Reakit System
 */

import { useMemo, useState } from "react";
import { useCompositeState } from "reakit";
import { useDateFormatter } from "@react-aria/i18n";
import { useControlledState } from "@react-stately/utils";

import { DateTimeFormatOptions } from "../utils/types";
import { add, setSegment, convertValue, getSegmentLimits } from "./__utils";

export interface IDateSegment {
  type: Intl.DateTimeFormatPartTypes;
  text: string;
  value?: number;
  minValue?: number;
  maxValue?: number;
  isPlaceholder: boolean;
}

const EDITABLE_SEGMENTS = {
  year: true,
  month: true,
  day: true,
  hour: true,
  minute: true,
  second: true,
  dayPeriod: true,
};

const PAGE_STEP = {
  year: 5,
  month: 2,
  day: 7,
  hour: 2,
  minute: 15,
  second: 15,
};

// Node seems to convert everything to lowercase...
const TYPE_MAPPING = {
  dayperiod: "dayPeriod",
};

export interface SegmentStateProps {
  value?: Date;
  defaultValue?: Date;
  formatOptions?: DateTimeFormatOptions;
  placeholderDate?: Date;
  onChange?: (value: Date, ...args: any[]) => void;
}

export function useSegmentState(props: SegmentStateProps) {
  const segmentComposite = useCompositeState({ orientation: "horizontal" });
  const [validSegments, setValidSegments] = useState(
    props.value || props.defaultValue ? { ...EDITABLE_SEGMENTS } : {},
  );

  // https://stackoverflow.com/a/9893752/10629172
  const dateFormatter = useDateFormatter(props.formatOptions);
  const resolvedOptions = useMemo(() => dateFormatter.resolvedOptions(), [
    dateFormatter,
  ]);

  // Determine how many editable segments there are for validation purposes.
  // The result is cached for performance.
  const numSegments = useMemo(
    () =>
      dateFormatter
        .formatToParts(new Date())
        .filter(seg => EDITABLE_SEGMENTS[seg.type]).length,
    [dateFormatter],
  );

  // If there is a value prop, and some segments were previously placeholders, mark them all as valid.
  if (props.value && Object.keys(validSegments).length < numSegments) {
    setValidSegments({ ...EDITABLE_SEGMENTS });
  }

  // We keep track of the placeholder date separately in state so that onChange is not called
  // until all segments are set. If the value === null (not undefined), then assume the component
  // is controlled, so use the placeholder as the value until all segments are entered so it doesn't
  // change from uncontrolled to controlled and emit a warning.
  const [placeholderDate, setPlaceholderDate] = useState(
    convertValue(props.placeholderDate) ||
      new Date(new Date().getFullYear(), 0, 1),
  );
  const [date, setDate] = useControlledState<Date>(
    // @ts-ignore
    props.value === null
      ? convertValue(placeholderDate)
      : convertValue(props.value),
    convertValue(props.defaultValue),
    props.onChange,
  );

  // If all segments are valid, use the date from state, otherwise use the placeholder date.
  const value =
    Object.keys(validSegments).length >= numSegments ? date : placeholderDate;
  const setValue = (value: Date) => {
    if (Object.keys(validSegments).length >= numSegments) {
      setDate(value);
    } else {
      setPlaceholderDate(value);
    }
  };

  const segments = dateFormatter.formatToParts(value).map(
    segment =>
      ({
        type: TYPE_MAPPING[segment.type] || segment.type,
        text: segment.value,
        ...getSegmentLimits(value, segment.type, resolvedOptions),
        isPlaceholder: !validSegments[segment.type],
      } as IDateSegment),
  );

  const adjustSegment = (
    type: Intl.DateTimeFormatPartTypes,
    amount: number,
  ) => {
    validSegments[type] = true;
    setValidSegments({ ...validSegments });
    // @ts-ignore
    setValue(add(value, type, amount, resolvedOptions));
  };

  return {
    ...segmentComposite,
    fieldValue: value,
    setFieldValue: setValue,
    segments,
    dateFormatter,
    increment(part: Intl.DateTimeFormatPartTypes) {
      adjustSegment(part, 1);
    },
    decrement(part: Intl.DateTimeFormatPartTypes) {
      adjustSegment(part, -1);
    },
    incrementPage(part: Intl.DateTimeFormatPartTypes) {
      adjustSegment(part, PAGE_STEP[part] || 1);
    },
    decrementPage(part: Intl.DateTimeFormatPartTypes) {
      adjustSegment(part, -(PAGE_STEP[part] || 1));
    },
    setSegment(part: Intl.DateTimeFormatPartTypes, v: number) {
      validSegments[part] = true;
      setValidSegments({ ...validSegments });
      // @ts-ignore
      setValue(setSegment(value, part, v, resolvedOptions));
    },
    confirmPlaceholder(part: Intl.DateTimeFormatPartTypes) {
      validSegments[part] = true;
      setValidSegments({ ...validSegments });
      setValue(new Date(value));
    },
  };
}

export type SegmentStateReturn = ReturnType<typeof useSegmentState>;
