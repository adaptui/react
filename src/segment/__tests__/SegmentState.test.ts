import { act, renderHook } from "reakit-test-utils/hooks";

import { useSegmentState, SegmentStateProps } from "../SegmentState";

function render({ ...initialState }: SegmentStateProps = {}) {
  return renderHook(() => useSegmentState({ ...initialState })).result;
}

describe("SegmentState", () => {
  it("should have proper segments", () => {
    const result = render({
      defaultValue: new Date(2050, 0, 0),
    });

    expect(result.current.segments).toMatchInlineSnapshot(`
      Array [
        Object {
          "isPlaceholder": false,
          "maxValue": 12,
          "minValue": 1,
          "text": "12",
          "type": "month",
          "value": 12,
        },
        Object {
          "isPlaceholder": true,
          "text": "/",
          "type": "literal",
        },
        Object {
          "isPlaceholder": false,
          "maxValue": 31,
          "minValue": 1,
          "text": "31",
          "type": "day",
          "value": 31,
        },
        Object {
          "isPlaceholder": true,
          "text": "/",
          "type": "literal",
        },
        Object {
          "isPlaceholder": false,
          "maxValue": 9999,
          "minValue": 1,
          "text": "2049",
          "type": "year",
          "value": 2049,
        },
      ]
    `);
  });

  it.each(["increment", "decrement"])(
    "should be able to %s a segment",
    type => {
      const result = render();

      expect(result.current.segments[0].value).toBe(1);
      act(() => {
        result.current[type]("month");
      });
      expect(result.current.segments[0].value).toBe(
        type === "increment" ? 2 : 12,
      );

      expect(result.current.segments[2].value).toBe(1);
      act(() => {
        result.current[type]("day");
      });

      expect(result.current.segments[2].value).toBe(
        type === "increment" ? 2 : 31,
      );
    },
  );

  it.each(["incrementPage", "decrementPage"])(
    "should be able to %s a segment",
    type => {
      const result = render();

      expect(result.current.segments[0].value).toBe(1);
      act(() => {
        result.current[type]("month");
      });
      expect(result.current.segments[0].value).toBe(
        type === "incrementPage" ? 3 : 11,
      );

      expect(result.current.segments[2].value).toBe(1);
      act(() => {
        result.current[type]("day");
      });
      expect(result.current.segments[2].value).toBe(
        type === "incrementPage" ? 8 : 24,
      );
    },
  );

  it("should be able to setSegment", () => {
    const result = render();

    expect(result.current.segments[0].value).toBe(1);

    act(() => {
      result.current.setSegment("month", 8);
    });
    expect(result.current.segments[0].value).toBe(8);

    act(() => {
      result.current.setSegment("day", 15);
    });
    expect(result.current.segments[2].value).toBe(15);

    act(() => {
      result.current.setSegment("year", 2080);
    });
    expect(result.current.segments[4].value).toBe(2080);
  });

  it("should be able to setFieldValue", () => {
    const result = render({ defaultValue: new Date(2050, 0, 0) });

    expect(result.current.fieldValue).toStrictEqual(new Date(2050, 0, 0));

    act(() => {
      result.current.setFieldValue(new Date(2030, 5, 5));
    });
    expect(result.current.fieldValue).toStrictEqual(new Date(2030, 5, 5));
  });

  it("should support different formatOptions", () => {
    const result = render({
      defaultValue: new Date(2050, 0, 0),
      formatOptions: {
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
        weekday: "long",
      },
    });

    expect(result.current.segments).toMatchInlineSnapshot(`
      Array [
        Object {
          "isPlaceholder": true,
          "text": "Friday",
          "type": "weekday",
        },
        Object {
          "isPlaceholder": true,
          "text": ", ",
          "type": "literal",
        },
        Object {
          "isPlaceholder": false,
          "maxValue": 12,
          "minValue": 1,
          "text": "12",
          "type": "month",
          "value": 12,
        },
        Object {
          "isPlaceholder": true,
          "text": "/",
          "type": "literal",
        },
        Object {
          "isPlaceholder": false,
          "maxValue": 31,
          "minValue": 1,
          "text": "31",
          "type": "day",
          "value": 31,
        },
        Object {
          "isPlaceholder": true,
          "text": "/",
          "type": "literal",
        },
        Object {
          "isPlaceholder": false,
          "maxValue": 9999,
          "minValue": 1,
          "text": "2049",
          "type": "year",
          "value": 2049,
        },
      ]
    `);
  });
});
