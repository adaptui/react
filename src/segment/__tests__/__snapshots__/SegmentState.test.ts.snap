// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentState should have proper segments 1`] = `
Array [
  Object {
    "isPlaceholder": false,
    "maxValue": 12,
    "minValue": 1,
    "text": "12",
    "type": "month",
    "value": 12,
  },
  Object {
    "isPlaceholder": true,
    "text": "/",
    "type": "literal",
  },
  Object {
    "isPlaceholder": false,
    "maxValue": 31,
    "minValue": 1,
    "text": "31",
    "type": "day",
    "value": 31,
  },
  Object {
    "isPlaceholder": true,
    "text": "/",
    "type": "literal",
  },
  Object {
    "isPlaceholder": false,
    "maxValue": 9999,
    "minValue": 1,
    "text": "2049",
    "type": "year",
    "value": 2049,
  },
]
`;

exports[`SegmentState should support different formatOptions 1`] = `
Array [
  Object {
    "isPlaceholder": true,
    "text": "Friday",
    "type": "weekday",
  },
  Object {
    "isPlaceholder": true,
    "text": ", ",
    "type": "literal",
  },
  Object {
    "isPlaceholder": false,
    "maxValue": 12,
    "minValue": 1,
    "text": "12",
    "type": "month",
    "value": 12,
  },
  Object {
    "isPlaceholder": true,
    "text": "/",
    "type": "literal",
  },
  Object {
    "isPlaceholder": false,
    "maxValue": 31,
    "minValue": 1,
    "text": "31",
    "type": "day",
    "value": 31,
  },
  Object {
    "isPlaceholder": true,
    "text": "/",
    "type": "literal",
  },
  Object {
    "isPlaceholder": false,
    "maxValue": 9999,
    "minValue": 1,
    "text": "2049",
    "type": "year",
    "value": 2049,
  },
]
`;
